Bluetooth: Audio: samples: Add multi subgroup support

jthm-ot opened the pull request:
Add multi subgroup support for the broadcast sink sample

jthm-ot committed:
Bluetooth samples: Add multi subgroup support

Add multi subgroup support for the broadcast sink sample

Signed-off-by: Jens Rehhoff Thomsen <jthm@demant.com>

kruithofa commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -732,89 +677,116 @@ static bool find_valid_bis_in_subgroup_bis(const struct bt_bap_base_subgroup *su
-			 * satisfy CONFIG_TARGET_BROADCAST_CHANNEL.
-			 */
+		for (uint8_t i = 0U; i < data.cnt; i++) {
+			bis_chan_allocation |= data.bis[i].chan_allocation;
```
```suggestion
			chan_allocation |= data.bis[i].chan_allocation;
```

jthm-ot replied:
No chan_allocation will be set in code below based on subgroup / bis level choice

kruithofa commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -732,89 +677,116 @@ static bool find_valid_bis_in_subgroup_bis(const struct bt_bap_base_subgroup *su
+		}
+
+		printk("Channel allocation (bis level) 0x%x\n", bis_chan_allocation);
+	}
```
```suggestion
	
```

jthm-ot replied:
Ok. The comment not needed. I will remove it

jthm-ot replied:
Ok

kruithofa commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -732,89 +677,116 @@ static bool find_valid_bis_in_subgroup_bis(const struct bt_bap_base_subgroup *su
-			if (find_valid_bis_in_subgroup_bis(subgroup, bis_indexes)) {
-				/* Found BISes with correct channel allocation */
-				return false;
+	if (bis_chan_allocation_available) {
```
I think, together with the previous suggestion, we do not need the if -statement
```suggestion

```

(github doesn't allow me to combine these 2 suggestions)

jthm-ot replied:
Ok

kruithofa commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -606,118 +608,61 @@ static struct bt_bap_stream_ops stream_ops = {
- * It returns `true` if the BIS either contains none or some of the channels we are looking for.
- * If it contains some, then that is being stored in the user_data, so that the calling function
- * can check if a combination of the BIS satisfy the channel allocations we want.
+ * Gets BIS channel allocation (if exists) and updates the bis_channel_allocation_data struct
+ * Returns `true` to continue to next BIS
```
Nitpick, but I find this slightly clearer
```suggestion
 * Gets BIS channel allocation (if it exists) and updates the bis_channel_allocation_data struct
 * Always returns `true` to continue to next BIS
```

jthm-ot replied:
Fine with the nitpick :-)

Thalley commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -606,118 +608,61 @@ static struct bt_bap_stream_ops stream_ops = {
+		return true; /* continue to next BIS */
 	}
 
+	/* bt_bap_base_subgroup_bis_codec_to_codec_cfg only updates the codec specific capabilities
+	 * data. The bap_base_codec_id data in the codec_cfg struct needed by
+	 * bt_audio_codec_cfg_get_chan_allocation for the fallback check.
+	 */
+	codec_cfg.id = data->bap_base_codec_id.id;
+	codec_cfg.cid = data->bap_base_codec_id.cid;
+	codec_cfg.vid = data->bap_base_codec_id.vid;
```
Furthermore the `codec_cfg` of BIS does not contain these values at all. When `bt_bap_base_subgroup_bis_codec_to_codec_cfg` is used at level 3, only the `codec_cfg.data` and `codec_cfg.data_len` contains valid values (as described by the documentation)

Thalley commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -732,89 +677,116 @@ static bool find_valid_bis_in_subgroup_bis(const struct bt_bap_base_subgroup *su
+static bool subgroup_get_valid_bis_indexes_cb(const struct bt_bap_base_subgroup *subgroup,
+					      void *user_data)
 {
+	enum bt_audio_location subgroup_chan_allocation = BT_AUDIO_LOCATION_MONO_AUDIO;
+	enum bt_audio_location bis_chan_allocation = BT_AUDIO_LOCATION_MONO_AUDIO;
```
Instead of instantiating these values, you can actually just call `bt_audio_codec_cfg_get_chan_allocation` with `fallback` to `true` which has the same behavior

Thalley commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -732,89 +677,116 @@ static bool find_valid_bis_in_subgroup_bis(const struct bt_bap_base_subgroup *su
-				 */
-				return true;
-			}
+	/* Keep base code id (used in bt_audio_codec_cfg_get_chan_allocation at bis level) */
+	struct bis_channel_allocation_data data = {.cnt = 0,
+						   .bap_base_codec_id = {.id = codec_cfg.id,
+									 .cid = codec_cfg.cid,
+									 .vid = codec_cfg.vid}};
```
```suggestion
	/* Keep base code id (used in bt_audio_codec_cfg_get_chan_allocation at bis level) */
	struct bis_channel_allocation_data data = {.cnt = 0,
						   .bap_base_codec_id = {.id = codec_cfg.id,
									 .cid = codec_cfg.cid,
									 .vid = codec_cfg.vid,},};
```

If you use clang-format, please end structs with a final `,` so that it formats it correctly instead of these weird formats :D

Thalley replied:
Kind of moot, as this struct is pretty useless as described above :)

Thalley commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -732,89 +677,116 @@ static bool find_valid_bis_in_subgroup_bis(const struct bt_bap_base_subgroup *su
-			*bis_indexes = BIT(find_lsb_set(subgroup_bis_indexes) - 1);
+	/* Get channel allocation at BIS level */
+	err = bt_bap_base_subgroup_foreach_bis(subgroup, bis_get_channel_allocation_cb, &data);
+	if (err == 0 && data.cnt > 0) {
+		bis_chan_allocation_available = true;
```
Is this correct? `if (err == 0 && data.cnt > 0)` can be true without the BIS having a channel allocation. 

E.g. if we have 

Subgroup[0] with chan_alloc = LEFT
BIS[0] with chan_alloc = RIGHT
BIS[1] without chan alloc

Then we set `bis_chan_allocation_available` to `true` and assume that all BIS has a channel allocation, which isn't always the case

jthm-ot replied:
Right. I will make the bis_get_channel_allocation_cb return false on bis channel allocation, failure then "if (err == 0)" should be the check to use

Thalley commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -732,89 +677,116 @@ static bool find_valid_bis_in_subgroup_bis(const struct bt_bap_base_subgroup *su
+					      void *user_data)
 {
+	enum bt_audio_location subgroup_chan_allocation = BT_AUDIO_LOCATION_MONO_AUDIO;
+	enum bt_audio_location bis_chan_allocation = BT_AUDIO_LOCATION_MONO_AUDIO;
```
The concept of having a single `bis_chan_allocation` for the subgroup is kind of weird. 

For example later we have
```
chan_allocation = bis_chan_allocation;
if (chan_allocation == CONFIG_TARGET_BROADCAST_CHANNEL) {
```

But if `CONFIG_TARGET_BROADCAST_CHANNEL` == `LEFT | RIGHT` and `bis_chan_allocation == LEFT | RIGHT | FRONT_CENTER` when number of BIS are `3`, then the check fails even if the subgroup has a LEFT and RIGHT bis that we are looking for

jthm-ot replied:
Right. I will change the code

Thalley commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -732,89 +677,116 @@ static bool find_valid_bis_in_subgroup_bis(const struct bt_bap_base_subgroup *su
+		chan_allocation = bis_chan_allocation;
+		/* Get the BIS indexes at BIS level */
+		for (uint8_t i = 0U; i < data.cnt; i++) {
+			if (data.bis[i].chan_allocation == CONFIG_TARGET_BROADCAST_CHANNEL ||
+			    (data.bis[i].chan_allocation & CONFIG_TARGET_BROADCAST_CHANNEL) != 0) {
```
Do we need both `==` and `&`? Wont `&` always work here?

jthm-ot replied:
No. Channel allocation as well as targer could be MONO (= 0)

Thalley commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -831,16 +803,26 @@ static void base_recv_cb(struct bt_bap_broadcast_sink *sink, const struct bt_bap
 	printk("Received BASE with %d subgroups from broadcast sink %p\n",
 	       bt_bap_base_get_subgroup_count(base), sink);
 
+	/**
+	 * Get a 32-bit bitfield of BIS indexes that cover the channel allocation values in
+	 * CONFIG_TARGET_BROADCAST_CHANNEL.
+	 */
```
This comment should be moved to the `#if defined(CONFIG_TARGET_BROADCAST_CHANNEL)` clause

larsgk commented on samples/bluetooth/bap_broadcast_sink/prj.conf:
```diff
@@ -6,14 +6,20 @@ CONFIG_BT_PAC_SNK=y
+CONFIG_BT_BAP_BASS_MAX_SUBGROUPS=4
 CONFIG_BT_BUF_ACL_RX_SIZE=255
 CONFIG_BT_BUF_ACL_TX_SIZE=251
+CONFIG_BT_AUDIO_CODEC_CFG_MAX_METADATA_SIZE=64
+CONFIG_BT_AUDIO_CODEC_CFG_MAX_DATA_SIZE=64
+CONFIG_BT_AUDIO_CODEC_CAP_MAX_METADATA_SIZE=64
+CONFIG_BT_AUDIO_CODEC_CAP_MAX_DATA_SIZE=64
```
We have seen sources (in market) that easily break the boundaries - requiring these values to be bumped.  Maybe we should even consider to default to theoretical max and let the application conf decrease the values to something lower if it is certain that the sink will never encounter something with large values.

Thalley replied:
Yeah, it makes sense that we use above minimum values, but it's still unrelated to the subgroup change in this PR. It is OK that they are in this PR, but please move to a separate commit explaining the reason for this change.

jthm-ot replied:
Ok

jthm-ot commented on samples/bluetooth/bap_broadcast_sink/prj.conf:
```diff
@@ -6,14 +6,20 @@ CONFIG_BT_PAC_SNK=y
 # Allocate 2 RX buffers per channel
-CONFIG_BT_ISO_RX_BUF_COUNT=4
-CONFIG_BT_BAP_BROADCAST_SNK_SUBGROUP_COUNT=2
-CONFIG_BT_BAP_BROADCAST_SNK_STREAM_COUNT=2
-CONFIG_BT_BAP_BASS_MAX_SUBGROUPS=2
+CONFIG_BT_ISO_RX_BUF_COUNT=8
+CONFIG_BT_BAP_BROADCAST_SNK_SUBGROUP_COUNT=4
+CONFIG_BT_BAP_BROADCAST_SNK_STREAM_COUNT=4
+CONFIG_BT_BAP_BASS_MAX_SUBGROUPS=4
```
I have just used more subgroups for test purpose. No need to increase in this PR. I will bring it back to 2

jthm-ot commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -143,13 +143,13 @@ static const struct bt_audio_codec_cap codec_cap = BT_AUDIO_CODEC_CAP_LC3(
- * 0.
+ * we have. Bit 0 is BIS index 1.
  */
-static const uint32_t bis_index_mask = BIT_MASK(ARRAY_SIZE(streams) + 1U);
+static const uint32_t bis_index_mask = BIT_MASK(ARRAY_SIZE(streams));
```
Correct. But since its just a sample I will include the fix in this commit

jthm-ot commented on samples/bluetooth/bap_broadcast_sink/prj.conf:
```diff
@@ -6,14 +6,20 @@ CONFIG_BT_PAC_SNK=y
+CONFIG_BT_AUDIO_CODEC_CAP_MAX_METADATA_SIZE=64
+CONFIG_BT_AUDIO_CODEC_CAP_MAX_DATA_SIZE=64
+
+CONFIG_BT_RX_STACK_SIZE=6144
```
No reason for the very specific value :-) More stack needed when more subgroups and more meta data included. I will decrease the number of subgroups to the previous value (2). Then the previous stack value should also be fine

Thalley replied:
Yeah the metadata size will affect this if we stack allocate a `codec_cfg` struct

Thalley commented:
@jthm-ot as per https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#workflow-suggestions-that-help-reviewers please do not resolve comments unless they are applied exactly :)

Thalley commented on samples/bluetooth/bap_broadcast_sink/prj.conf:
```diff
@@ -14,7 +14,12 @@ CONFIG_BT_BAP_BROADCAST_SNK_STREAM_COUNT=2
 CONFIG_BT_DEVICE_NAME="Broadcast Audio Sink"
 
 CONFIG_BT_TINYCRYPT_ECC=y
+
```
```suggestion
```

Thalley commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -143,13 +143,13 @@ static const struct bt_audio_codec_cap codec_cap = BT_AUDIO_CODEC_CAP_LC3(
 static uint32_t requested_bis_sync;
 static uint32_t bis_index_bitfield;
 static uint8_t sink_broadcast_code[BT_AUDIO_BROADCAST_CODE_SIZE];
+static uint32_t requested_subgroup_sync;
```
Still not clear that `requested_subgroup_sync` is a bitfield.

Since this is a sample, we should document it as well

jthm-ot replied:
I will add a comment

Thalley commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -606,118 +608,66 @@ static struct bt_bap_stream_ops stream_ops = {
 	struct {
 		uint8_t index;
 		enum bt_audio_location chan_allocation;
+		bool chan_allocation_available;
```
Suggest to move the bool above `chan_allocation` or `index` below `chan_allocation` so we don't have `1 | 4 | 1` in memory

jthm-ot replied:
Ok. Good point

Thalley commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -606,118 +608,66 @@ static struct bt_bap_stream_ops stream_ops = {
 	if (err != 0) {
 		printk("Could not get codec configuration for BIS: %d\n", err);
-		return true;
+		data->bis[idx].chan_allocation_available = false;
```
Suggest to move `data->bis[idx].chan_allocation_available = false;` to line 638 so it's initialized to `false` and then only update it to `true` when setting it

jthm-ot replied:
Will do

Thalley commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -606,118 +608,66 @@ static struct bt_bap_stream_ops stream_ops = {
 	err = bt_audio_codec_cfg_get_chan_allocation(&codec_cfg, &chan_allocation, true);
 	if (err != 0) {
 		printk("Could not find channel allocation for BIS: %d\n", err);
+		data->bis[idx].chan_allocation_available = false;
```
```suggestion
```

See above comment

Thalley commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -606,118 +608,66 @@ static struct bt_bap_stream_ops stream_ops = {
 
-	return true;
-}
+	/* Store channel allocation data for this bis */
+	data->bis[idx].chan_allocation = chan_allocation;
+	data->bis[idx].chan_allocation_available = true;
```
What does `chan_allocation_available` represent here? 

Since you are using fallback above, then this will be set to `true` regardless of whether there's a location bit set for the BIS.

jthm-ot replied:
Correct. But if fallback is not used this will ensure that subgroup allocation is used for this BIS. Also the case if bt_bap_base_subgroup_bis_codec_to_codec_cfg somehow fails

Thalley commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -732,89 +682,109 @@ static bool find_valid_bis_in_subgroup_bis(const struct bt_bap_base_subgroup *su
-			 * We continue searching each BIS to get the minimal amount of BIS that
-			 * satisfy CONFIG_TARGET_BROADCAST_CHANNEL.
-			 */
+	if (chan_allocation == CONFIG_TARGET_BROADCAST_CHANNEL) {
```
`chan_allocation` is always `BT_AUDIO_LOCATION_MONO_AUDIO`. It is never updated

Thalley commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -732,89 +682,107 @@ static bool find_valid_bis_in_subgroup_bis(const struct bt_bap_base_subgroup *su
-				 */
-				return true;
-			}
+	/* Get the BIS indexes */
+	for (uint8_t i = 0U; i < data.cnt; i++) {
+		if (data.bis[i].chan_allocation == CONFIG_TARGET_BROADCAST_CHANNEL ||
+		    (data.bis[i].chan_allocation & CONFIG_TARGET_BROADCAST_CHANNEL) != 0) {
+			bis_indexes |= BT_ISO_BIS_INDEX_BIT(data.bis[i].index);
+		}
+	}
```
If `CONFIG_TARGET_BROADCAST_CHANNEL` == LEFT | RIGHT and `data.bis[0].chan_allocation` is LEFT, `data.bis[1].chan_allocation` is RIGHT and `data.bis[2].chan_allocation` is LEFT | RIGHT then `bis_indexes` becomes `BIT(0) | BIT(1) | BIT(2)` which is not what we want. In this case we want only either `BIT(0) | BIT(1)` or `BIT(2)`.

Thalley reviewed (changes requested):
I think you are trying to modify too many things here. 

Increasing the number of subgroups should not require modifying the BIS index selection really, as we probably never want to sync to BIS from different subgroups anyways.

I suggest that you focus this PR on the subgroup count support and then if you want to refactor the BIS selection, then do that in a separate PR (or at least separate commit).

jthm-ot commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -555,6 +556,7 @@ static void stream_stopped_cb(struct bt_bap_stream *stream, uint8_t reason)
 	}
 
 	if (k_sem_count_get(&sem_stream_started) != stream_count) {
+		printk("BIG not synced\n");
```
Ok :-)

jthm-ot committed:
Bluetooth samples: Increase codec meta data size

To improve broadcast sink sample robostness the max codec meta data
size configuration is increased to reflect the meta codec data sizes
seen for sources in market.

Signed-off-by: Jens Rehhoff Thomsen <jthm@demant.com>

jthm-ot commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -393,7 +393,7 @@ static int lc3_enable(struct broadcast_sink_stream *sink_stream)
 	sdu_size_required = chan_alloc_bit_cnt * sink_stream->lc3_octets_per_frame *
 			    sink_stream->lc3_frames_blocks_per_sdu;
-	if (sdu_size_required < sink_stream->stream.qos->sdu) {
+	if (sdu_size_required > sink_stream->stream.qos->sdu) {
```
Ok. I will make a bug fix commit and include it in this PR

jthm-ot committed:
Bluetooth samples: Fix wrong SDU size check

Just a typo in broadcast sink sample main code.
'<' changed to '>' in if statement.

Signed-off-by: Jens Rehhoff Thomsen <jthm@demant.com>

jthm-ot committed:
Bluetooth samples: Fix BIS index mask variable

For the BIS index mask variable used in broadcast sink sample
bit 0 is now BIS index 1.

Signed-off-by: Jens Rehhoff Thomsen <jthm@demant.com>

jthm-ot commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -606,118 +608,61 @@ static struct bt_bap_stream_ops stream_ops = {
- * It returns `true` if the BIS either contains none or some of the channels we are looking for.
- * If it contains some, then that is being stored in the user_data, so that the calling function
- * can check if a combination of the BIS satisfy the channel allocations we want.
+ * Gets BIS channel allocation (if exists) and updates the bis_channel_allocation_data struct
```
I guess "Gets BIS channel allocation (if exists)" should be enough info. I will remove til "struct" part

Thalley commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -832,15 +808,24 @@ static void base_recv_cb(struct bt_bap_broadcast_sink *sink, const struct bt_bap
 		return;
 	}
 #else
-	err = bt_bap_base_get_bis_indexes(base, &base_bis_index_bitfield);
-	if (err != 0) {
-		printk("Failed to BIS indexes: %d\n", err);
```
Why was this changed? It seems to do the same

Thalley commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -1033,15 +1019,31 @@ static int bis_sync_req_cb(struct bt_conn *conn,
+
+	for (uint8_t subgroup = 0; subgroup < recv_state->num_subgroups; subgroup++) {
+		if (bis_sync_req[subgroup] != 0) {
+			/* bis_sync_req[subgroup] = 0b1 or BT_BAP_BIS_SYNC_NO_PREF */
```
```suggestion
```

Not sure what the point of this comment is. It's obvious that `bis_sync_req[subgroup]` is not 0 from the code `if (bis_sync_req[subgroup] != 0) {`

jthm-ot replied:
Yes. Comment not needed. Will delete it

Thalley commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -1033,15 +1019,31 @@ static int bis_sync_req_cb(struct bt_conn *conn,
+			if (requested_bis_sync == 0) {
+				requested_bis_sync = bis_sync_req[subgroup];
+			} else {
+				/* requested_bis_sync = 0b1 or BT_BAP_BIS_SYNC_NO_PREF */
```
```suggestion
```

Ditto

jthm-ot replied:
Yes. Comment not needed. Will delete it

Thalley commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -1033,15 +1019,31 @@ static int bis_sync_req_cb(struct bt_conn *conn,
+				requested_bis_sync = bis_sync_req[subgroup];
+			} else {
+				/* requested_bis_sync = 0b1 or BT_BAP_BIS_SYNC_NO_PREF */
+				if (requested_bis_sync != BT_BAP_BIS_SYNC_NO_PREF &&
+				    bis_sync_req[subgroup] != BT_BAP_BIS_SYNC_NO_PREF) {
+					printk("Bis sync request from more than 1 subgroup!\n");
+					return -EINVAL;
+				}
```
Suggest to add a comment why we are rejecting this. Why are we? Doesn't this prevent us from syncing to multiple subgroups?

Thalley replied:
> request from more than 1 subgroup is not possible unless sync value is 0 or BT_BAP_BIS_SYNC_NO_PREF.

Can you elaborate?
From a spec perspektive it's possible to sync to BIS 1 from Subgroup 1 and BIS 2 from Subgroup 2.

larsgk replied:
> > request from more than 1 subgroup is not possible unless sync value is 0 or BT_BAP_BIS_SYNC_NO_PREF.
> 
> Can you elaborate? From a spec perspektive it's possible to sync to BIS 1 from Subgroup 1 and BIS 2 from Subgroup 2.

@Thalley  - would this make much sense in practice? (multi subgroup sync)

Thalley replied:
I don't have a good use case for it to be honest :) But it's incorrect to say that it's not possible, and I wouldn't be against not supporting it in the sample, and if we do not support syncing to multiple subgroups then we remove `requested_subgroup_sync` again as well as add a comment why this isn't implemented.

jthm-ot replied:
When reading "6.5.4. Adding broadcast sources" from BAP spec
 
"The Broadcast Assistant shall not write a value of 0b1 to any BIS_Sync parameter BIS_index value for more than one subgroup except if the Broadcast Assistant writes a value of 0xFFFFFFFF (no preference) to the BIS_Sync parameter for each of those subgroups."

You can be i doubt about sync request from multiple subgroups is possible. I prefer to leave it as is an make a comment.

Also it would be a strange use case to stream from multiple subgroups  at the same time :-)

Thalley commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -1033,15 +1019,31 @@ static int bis_sync_req_cb(struct bt_conn *conn,
+		}
+	}
+
+	printk("BIS sync req for %p: 0x%08x (subgroup mask 0x%08x), broadcast id: 0x%06x, (%s)\n",
+	       recv_state, requested_bis_sync, requested_subgroup_sync, recv_state->broadcast_id,
```
```suggestion
	printk("BIS sync req for %p: BIS indexes 0x%08x (subgroup indexes 0x%08x), broadcast id: 0x%06x, (%s)\n",
	       recv_state, requested_bis_sync, requested_subgroup_sync, recv_state->broadcast_id,
```

jthm-ot replied:
Ok. Will use the suggestion

Thalley commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -1641,7 +1642,7 @@ int main(void)
 		sync_bitfield = bis_index_bitfield & requested_bis_sync;
 		stream_count = 0;
-		for (int i = 1; i < BT_ISO_MAX_GROUP_ISO_COUNT; i++) {
+		for (int i = 0; i < BT_ISO_MAX_GROUP_ISO_COUNT; i++) {
```
This is another bugfix not related to the subgroup change

jthm-ot replied:
Danm. Then I will put it in the "Fix BIS index mask variable" commit

jthm-ot commented on samples/bluetooth/bap_broadcast_sink/prj.conf:
```diff
@@ -14,6 +14,10 @@ CONFIG_BT_BAP_BROADCAST_SNK_STREAM_COUNT=2
 CONFIG_BT_BAP_BASS_MAX_SUBGROUPS=2
 CONFIG_BT_BUF_ACL_RX_SIZE=255
 CONFIG_BT_BUF_ACL_TX_SIZE=251
+CONFIG_BT_AUDIO_CODEC_CFG_MAX_METADATA_SIZE=64
+CONFIG_BT_AUDIO_CODEC_CFG_MAX_DATA_SIZE=64
+CONFIG_BT_AUDIO_CODEC_CAP_MAX_METADATA_SIZE=64
+CONFIG_BT_AUDIO_CODEC_CAP_MAX_DATA_SIZE=64
```
Ok. I will improve the commit message

jthm-ot commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -393,7 +393,7 @@ static int lc3_enable(struct broadcast_sink_stream *sink_stream)
 	sdu_size_required = chan_alloc_bit_cnt * sink_stream->lc3_octets_per_frame *
 			    sink_stream->lc3_frames_blocks_per_sdu;
-	if (sdu_size_required < sink_stream->stream.qos->sdu) {
+	if (sdu_size_required > sink_stream->stream.qos->sdu) {
```
Ok. I will improve the commit message

jthm-ot commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -555,6 +558,7 @@ static void stream_stopped_cb(struct bt_bap_stream *stream, uint8_t reason)
 	}
 
 	if (k_sem_count_get(&sem_stream_started) != stream_count) {
+		printk("BIG synced termination\n");
```
Yes. Will fix it

jthm-ot commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -606,118 +610,65 @@ static struct bt_bap_stream_ops stream_ops = {
-		}
-	}
+	/* Store channel allocation data for this bis */
+	data->bis[idx].chan_allocation = chan_allocation;
```
Good point. Will do

jthm-ot commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -732,89 +683,114 @@ static bool find_valid_bis_in_subgroup_bis(const struct bt_bap_base_subgroup *su
 {
-	enum bt_audio_location chan_allocation;
+	enum bt_audio_location subgroup_chan_allocation;
+	enum bt_audio_location chan_allocation = 0;
```
Will initialize variable to BT_AUDIO_LOCATION_MONO_AUDIO then

jthm-ot commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -732,89 +683,114 @@ static bool find_valid_bis_in_subgroup_bis(const struct bt_bap_base_subgroup *su
 	}
 
-	err = bt_audio_codec_cfg_get_chan_allocation(&codec_cfg, &chan_allocation, false);
+	struct bis_channel_allocation_data data;
```
Will do

jthm-ot commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -732,89 +683,114 @@ static bool find_valid_bis_in_subgroup_bis(const struct bt_bap_base_subgroup *su
 
-			/* We found the BIS index we want, stop parsing*/
-			*bis_indexes = BIT(find_lsb_set(subgroup_bis_indexes) - 1);
+			printk("Channel allocation match. BIS index 0x%x\n", bis_indexes);
```
Ok

jthm-ot commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -732,89 +683,114 @@ static bool find_valid_bis_in_subgroup_bis(const struct bt_bap_base_subgroup *su
+		} else if ((data.bis[i].chan_allocation != 0) &&
+			   ((data.bis[i].chan_allocation & CONFIG_TARGET_BROADCAST_CHANNEL) ==
+			    data.bis[i].chan_allocation)) {
+			/* Party match */
```
Yes :-)

jthm-ot commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -732,89 +683,114 @@ static bool find_valid_bis_in_subgroup_bis(const struct bt_bap_base_subgroup *su
-			 * satisfy CONFIG_TARGET_BROADCAST_CHANNEL.
-			 */
+			return false; /* bis index found */
+		} else if ((data.bis[i].chan_allocation != 0) &&
```
Ok

jthm-ot commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -732,89 +683,114 @@ static bool find_valid_bis_in_subgroup_bis(const struct bt_bap_base_subgroup *su
+	} else if (err == -ENODATA) {
+		printk("No channel allocation data in subgroup\n");
+	} else {
+		/* subgroup error */
+		return true; /* continue to next subgroup */
```
Good point. Will do

jthm-ot commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -732,89 +683,114 @@ static bool find_valid_bis_in_subgroup_bis(const struct bt_bap_base_subgroup *su
-			 */
+			return false; /* bis index found */
+		} else if ((data.bis[i].chan_allocation != 0) &&
+			   ((data.bis[i].chan_allocation & CONFIG_TARGET_BROADCAST_CHANNEL) ==
+			    data.bis[i].chan_allocation)) {
```
Yes

Thalley replied:
Why? 

If we are interested in the LEFT channel and there is a stereo streaming, we can sync to that BIS and then just discard the RIGHT channel data. As far as I can tell, this check prevents us from syncing to any broadcast source with stereo BIS if we only support a single channel, which isn't ideal

jthm-ot replied:
Yes its complicated with this channel allocation. I will make a more greedy approach

jthm-ot commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -606,118 +610,65 @@ static struct bt_bap_stream_ops stream_ops = {
+		return true; /* continue to next BIS */
 	}
 
+	/* bt_bap_base_subgroup_bis_codec_to_codec_cfg only updates the codec specific capabilities
+	 * data. The bap_base_codec_id data in the codec_cfg struct needed by
+	 * bt_audio_codec_cfg_get_chan_allocation for the fallback check.
+	 */
```
I will add clarification in the comment

Thalley replied:
IMO it would be better to check if the codec ID is LC3 in the BASE in `subgroup_get_valid_bis_indexes_cb` and then stop parsing the BASE if it is not. LC3 is the only codec we support, so parsing non-LC3 subgroups is moot. 

If we apply this check, then it it is always LC3 when we reach this point, and we can just hardcode the `data.bap_base_codec_id.id = BT_HCI_CODING_FORMAT_LC3`

jthm-ot replied:
Okay. I will do that :-)

Thalley commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -555,6 +558,7 @@ static void stream_stopped_cb(struct bt_bap_stream *stream, uint8_t reason)
 	}
 
 	if (k_sem_count_get(&sem_stream_started) != stream_count) {
+		printk("BIG sync termination\n");
```
```suggestion
		printk("BIG sync terminated\n");
```

Thalley commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -618,106 +623,45 @@ struct find_valid_bis_data {
 	}
 
-	err = bt_audio_codec_cfg_get_chan_allocation(&codec_cfg, &chan_allocation, true);
+	/* Only LC3 codec supported (assumed at this point) */
+	codec_cfg.id = BT_HCI_CODING_FORMAT_LC3;
```
```suggestion
	/* Only LC3 codec supported (assumed at this point). 
	 * It is required to set the codec_cfg.id to use the fallback behavior of bt_audio_codec_cfg_get_chan_allocation. 
	 */
	codec_cfg.id = BT_HCI_CODING_FORMAT_LC3;
```

We could just not use the fallback behavior; up to you :)

jthm-ot replied:
Yes. Code reduction. I will take that

Thalley commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -732,89 +676,93 @@ static bool find_valid_bis_in_subgroup_bis(const struct bt_bap_base_subgroup *su
+	if (err == 0) {
+		printk("Channel allocation (subgroup level) 0x%x\n", subgroup_chan_allocation);
+		subgroup_chan_allocation_available = true;
+	} else if (err == -ENODATA) {
+		printk("No channel allocation data in subgroup\n");
```
```suggestion
```

Since we have checked for LC3 and using fallback behavior, this cannot happen

jthm-ot replied:
Right. Well spotted!

Thalley reviewed (changes requested):
A few minor comments, otherwise OK

jthm-ot commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -143,14 +143,16 @@ static const struct bt_audio_codec_cap codec_cap = BT_AUDIO_CODEC_CAP_LC3(
 static uint32_t bis_index_bitfield;
 static uint8_t sink_broadcast_code[BT_AUDIO_BROADCAST_CODE_SIZE];
 
+/* Bit field indicating from which subgroup(s) BIS sync is requested */
+static uint32_t requested_subgroup_sync; /* currently only used for printout */
```
Ok

jthm-ot commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -1033,15 +983,35 @@ static int bis_sync_req_cb(struct bt_conn *conn,
+			if (requested_bis_sync == 0) {
+				requested_bis_sync = bis_sync_req[subgroup];
+			} else {
+				if (requested_bis_sync != BT_BAP_BIS_SYNC_NO_PREF &&
+				    bis_sync_req[subgroup] != BT_BAP_BIS_SYNC_NO_PREF) {
```
if (bis_sync_req[subgroup] != 0) {

is in the beginning of the for-loop

jthm-ot commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -732,89 +670,91 @@ static bool find_valid_bis_in_subgroup_bis(const struct bt_bap_base_subgroup *su
+			bis_indexes |= BT_ISO_BIS_INDEX_BIT(data.bis[i].index);
 
-	return 0;
+			if ((chan_allocation & CONFIG_TARGET_BROADCAST_CHANNEL) ==
+			    CONFIG_TARGET_BROADCAST_CHANNEL) {
+				printk("Channel allocation match. BIS index bitfield 0x%x\n",
+				       bis_indexes);
+				*(uint32_t *)user_data = bis_indexes;
+
+				return false; /* bis indexes found */
+			}
```
No then it hits the if statement in the for loop

		if ((data.bis[i].chan_allocation == CONFIG_TARGET_BROADCAST_CHANNEL) ||
		    ((data.bis[i].chan_allocation & CONFIG_TARGET_BROADCAST_CHANNEL) ==
		     CONFIG_TARGET_BROADCAST_CHANNEL)) {
			/* Exact match */
			bis_indexes = BT_ISO_BIS_INDEX_BIT(data.bis[i].index);

and choose BIS[1]

Thalley replied:
Ah, right :)

jthm-ot commented on samples/bluetooth/bap_broadcast_sink/src/main.c:
```diff
@@ -832,17 +772,26 @@ static void base_recv_cb(struct bt_bap_broadcast_sink *sink, const struct bt_bap
+	 * CONFIG_TARGET_BROADCAST_CHANNEL.
+	 */
+	printk("Target channel location: 0x%x\n", CONFIG_TARGET_BROADCAST_CHANNEL);
+	err = bt_bap_base_foreach_subgroup(base, subgroup_get_valid_bis_indexes_cb,
+					   &base_bis_index_bitfield);
+	if (err != 0 && err != -ECANCELED) {
+		printk("Failed to get valid BIS indexes: %d\n", err);
```
Good point. Well spotted. Will do that
